Aplib.Core/Action.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// Describes an action that can be executed and guarded.
+    /// </summary>
+    /// <typeparam name="TQuery">The type of the query of the action</typeparam>
+    public class Action<TQuery>
+    {
+        private TQuery? _storedResult;
+
+        /// <summary>
+        /// Gets or sets the effect of the action.
+        /// </summary>
+        public System.Action<TQuery> Effect { private get; set; }
+
+        /// <summary>
+        /// Gets a value indicating whether the action is actionable.
+        /// </summary>
+        public bool Actionable => Query!.Invoke() is not (false or null);
+
+        /// <summary>
+        /// Gets or sets the query of the action. Can return a value, which is stored and can be used in the effect.
+        /// </summary>
+        /// <remarks>A boolean value can also be used. If an object is used, use null to mark the action as unactionable.</remarks>
+        public Func<TQuery> Query { private get; set; }
```

Comment:
Query not having a nullable type means users HAVE to implement a query. Many actions wouldn't need a guard, however, so it seems weird a Query is required. Masybe Query has a confusing naming, and should be changed to Guard (and I think the Guard method should always implicitly be called when Execute is called, or some similar behaviour).